File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/DigestProvider.kt
package com.blockchaincommons.bccomponents

/**
 * A type that can provide a single unique digest characterizing its contents.
 *
 * Implementations produce a cryptographic [Digest] (SHA-256) that uniquely
 * identifies the content of the implementing type.
 *
 * Use cases include data integrity verification, content-addressable storage,
 * and comparing objects by content rather than identity.
 */
interface DigestProvider {
    /** Returns a digest that uniquely characterizes this object's content. */
    fun digest(): Digest
}

/**
 * Extension that lets any [ByteArray] act as a [DigestProvider] by hashing
 * its contents.
 */
fun ByteArray.toDigest(): Digest = Digest.fromImage(this)

File: python/bc-components/src/bc_components/_digest_provider.py
"""DigestProvider protocol."""

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, runtime_checkable

if TYPE_CHECKING:
    from ._digest import Digest


@runtime_checkable
class DigestProvider(Protocol):
    """A type that can provide a single unique digest characterizing its contents."""

    def digest(self) -> Digest: ...
